### PR TITLE
Fix the slave registry failed

### PR DESCRIPTION
**Pull request by lokikiller**
When using the file discovery method, if a slave address failed to registry itself when the scanning the configuration file at the first time (maybe the slave's docker service donot start), the slave will not be registry itself again, even if the slave's docker service started, unless remove the slave's address from the configuration file and add it again.

So, I changed the "addEngine" function, appending a return parameter "cluster.Engine", when the parameter is null, delegating the engine registered failed, and I add a function to loop to registry it every 30s, until it registry successfully.

Signed-off-by: Tingyi Hao haotingyi13@otcaix.iscas.ac.cn
Signed-off-by: Tingyi Hao <haotingyi13@otcaix.iscas.ac.cn>

### CODE REVIEW

File: cluster/swarm/cluster.go [hunk 1]
```diff
@@ -158,10 +163,39 @@ func (c *Cluster) hasEngineByAddr(addr string) bool {
...
+		//				_, engine = c.registEngine(addr)
+		//				if engine != nil {
+		//					return
+		//				}
+		//			}
+		//		}()
+		ticker := time.NewTicker(reAddEnginePeriod)
+		go func() {
+			for {
+				select {
```
Comment (vieux) [inline cluster/swarm/cluster.go]: This is going to loop like crazy use so much CPU...

File: cluster/swarm/cluster.go [hunk 2]
```diff
@@ -158,10 +163,39 @@ func (c *Cluster) hasEngineByAddr(addr string) bool {
 	return c.getEngineByAddr(addr) != nil
 }
 
-func (c *Cluster) addEngine(addr string) bool {
+func (c *Cluster) addEngine(addr string) {
+	res, engine := c.registEngine(addr)
+	if !res && engine == nil {
+		//		ticker := time.Tick(reAddEnginePeriod)
```
Comment (vieux) [inline cluster/swarm/cluster.go]: why keep the comments ?

Comment (lokikiller) [inline cluster/swarm/cluster.go]: @vieux I think the comments way maybe better, but I imitate the code in "discovery/file/file.go" from 83 line to 100 line, to maintain the coding style.

**Comment by aluzzardi** [comment]
I understand the issue but I'm not sure this is the right approach.

The reason we don't attempt anymore is because of this;
```go
 			added, removed := currentEntries.Diff(entries)
 			currentEntries = entries
```

Meaning that we won't try to connect to an `added` engine more than once.

Not sure about a better solution yet

**Comment by aluzzardi** [comment]
@lokikiller Could you take a look at #1486?

**Comment by amitshukla** [comment]
@lokikiller Can you please let us know if you are interested in pursuing this PR?

**Comment by lokikiller** [comment]
@amitshukla yeah, I am interested in pursuing this PR

**Pull request closed without merging**